credential/cargo-credential-1password/src/lib.rs

```diff
@@ -262,53 +255,54 @@ impl OnePasswordKeychain {
-                index_url
-            )
-            .into());
-        }
-    }
-
-    fn store(&self, index_url: &str, token: &str, name: Option<&str>) -> Result<(), Error> {
-        let session = self.signin()?;
-        // Check if an item already exists.
-        if let Some(id) = self.search(&session, index_url)? {
-            self.modify(&session, &id, token, name)
-        } else {
-            self.create(&session, index_url, token, name)
-        }
-    }
-
-    fn erase(&self, index_url: &str) -> Result<(), Error> {
-        let session = self.signin()?;
-        // Check if an item already exists.
-        if let Some(id) = self.search(&session, index_url)? {
-            self.delete(&session, &id)?;
-        } else {
-            eprintln!("not currently logged in to `{}`", index_url);
+impl Credential for OnePasswordCredential {
+    fn perform(
+        &self,
+        registry: &RegistryInfo,
+        action: &Action,
+        args: &[&str],
+    ) -> Result<CredentialResponse, Error> {
+        let op = OnePasswordKeychain::new(args)?;
+        match action {
+            Action::Get(_) => {
+                let session = op.signin()?;
+                if let Some(id) = op.search(&session, registry.index_url)? {
+                    op.get_token(&session, &id)
+                        .map(|token| CredentialResponse::Get {
+                            token,
+                            cache: CacheControl::Session,
+                            operation_independent: true,
+                        })
+                } else {
+                    Err(Error::NotFound)
+                }
+            }
+            Action::Login(options) => {
+                let session = op.signin()?;
+                // Check if an item already exists.
+                if let Some(id) = op.search(&session, registry.index_url)? {
+                    eprintln!("note: token already exists for `{}`", registry.index_url);
+                    let token = cargo_credential::read_token(options, registry)?;
+                    op.modify(&session, &id, token.as_deref(), None)?;
+                } else {
+                    let token = cargo_credential::read_token(options, registry)?;
+                    op.create(&session, registry.index_url, token.as_deref(), None)?;
+                }
+                Ok(CredentialResponse::Login)
+            }
+            Action::Logout => {
+                let session = op.signin()?;
+                // Check if an item already exists.
+                if let Some(id) = op.search(&session, registry.index_url)? {
+                    op.delete(&session, &id)?;
+                    Ok(CredentialResponse::Logout)
+                } else {
+                    Err(Error::NotFound)
+                }
+            }
+            _ => Err(Error::OperationNotSupported),
         }
-        Ok(())
     }
 }
-
-fn main() {
```

Comment:
In the future, how hard would it be to translate the 1password provider back to a standalone executable? We haven't decided if we are going to support these providers directly. One option was to migrate these to a separate repo (perhaps under my name) if we decide that rust-lang isn't going to support these. The maintenance burden isn't too bad, so I'm not sure if it matters (and we can document that we can drop providers in the future if we can't support it).

Reply:
It's very easy to translate back. Just need to change it back to a `bin` crate and add:
```
fn main() {
    cargo_credential::main(OnePasswordCredential);
}
```

I think I'd also prefer that 1password was a standalone provider that you could `cargo install`.